mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/engine/implementation/MutableBoardPiece.kt

```diff
@@ -0,0 +1,110 @@
+   *
+   * @param id the [PieceIdentifier] for the piece.
+   * @param rank the [Rank] for the piece.
+   * @param color the [Color] for the piece.
+   */
+  constructor(id: PieceIdentifier, rank: Rank, color: Color) : this(pack(id, rank, color))
+
+  /** Returns true if the [flag] is contained within the [mask]. */
+  private fun hasFlag(mask: Int, flag: Int): Boolean = (packed and mask) == flag
+
+  /** Returns true iff this [MutableBoardPiece] represents an absent piece. */
+  val isNone: Boolean
+    get() = (packed and PieceColorRankFlag) == None.packed
+
+  /**
+   * Returns the unique identifier for this piece amongst all the [MutableBoardPiece] of same [Rank]
+   * and same [Color] on the board.
+   */
+  val id: Int
+    get() = (packed ushr 4)
+
+  /** Returns the [Color] associated to this [MutableBoardPiece], if it's not [None]. */
+  val color: Color?
+    get() =
+        when {
+          isNone -> null
+          hasFlag(mask = PieceColorMask, flag = PieceColorBlackFlag) -> Black
+          else -> White
+        }
+
+  /** Returns the [Rank] associated to this [MutableBoardPiece], if it's not [None]. */
+  val rank: Rank?
+    get() =
+        when {
+          hasFlag(mask = PieceRankMask, flag = PieceRankBishopFlag) -> Bishop
+          hasFlag(mask = PieceRankMask, flag = PieceRankKingFlag) -> King
+          hasFlag(mask = PieceRankMask, flag = PieceRankKnightFlag) -> Knight
+          hasFlag(mask = PieceRankMask, flag = PieceRankPawnFlag) -> Pawn
+          hasFlag(mask = PieceRankMask, flag = PieceRankQueenFlag) -> Queen
+          hasFlag(mask = PieceRankMask, flag = PieceRankRookFlag) -> Rook
+          else -> null
+        }
+
+  companion object {
+
+    /** Unpacks the provided [Int] and returns the associated [MutableBoardPiece]. */
+    fun fromPacked(byte: Int): MutableBoardPiece {
+      return MutableBoardPiece(byte)
+    }
+
+    /** Returns the [Int] representation of the provided [piece]. */
+    fun toInt(piece: MutableBoardPiece): Int {
+      return piece.packed
+    }
+
+    /** Returns a [MutableBoardPiece] which represents an empty cell. */
+    val None = MutableBoardPiece(0b0) // no rank
+  }
+}
+
+/** Packs the piece identifier, the [Rank] and the [Color] in an [Int]. */
+private fun pack(
+    id: PieceIdentifier,
+    rank: Rank,
+    color: Color,
+): Int = requireNotNull(ColorMap[color]) or requireNotNull(RankMap[rank]) or id.value.shl(4)
+
+// Masks and flags for the piece colors.
+private const val PieceColorMask = 0b1
+private const val PieceColorBlackFlag = 0b0
+private const val PieceColorWhiteFlag = 0b1
+
+// Masks and flags for the piece ranks.
+private const val PieceRankMask = 0b1110
+private const val PieceRankBishopFlag = 0b0010
+private const val PieceRankKingFlag = 0b0100
+private const val PieceRankKnightFlag = 0b0110
+private const val PieceRankPawnFlag = 0b1000
+private const val PieceRankQueenFlag = 0b1010
+private const val PieceRankRookFlag = 0b1110
```

Comment:
It actually is a part of the optimisation work, yes. By making `Piece` inline and by representing all the data within a single `Int` using some binary flags for the rank/color and the piece identifier, calls to `get` and `set` some pieces are compiled to passing and reading a primitive `Int`. This is also the reason for which `Piece.None` is returned on missing pieces, rather than a `Piece`, since it avoids primitive auto-boxing.

These flags are however well encapsulated in the `Piece` class, so it doesn't affect client code which uses this class.